include/zephyr/senss/senss_sensor.h

```diff
@@ -0,0 +1,594 @@
+/*
+ * Copyright (c) 2022-2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SENSS_SENSOR_H_
+#define ZEPHYR_INCLUDE_SENSS_SENSOR_H_
+
+#include <zephyr/senss/senss.h>
+#include <zephyr/device.h>
+#include <stdbool.h>
+#include <zephyr/drivers/sensor.h>
+
+/**
+ * @defgroup senss_sensor Sensor API
+ * @ingroup senss
+ * @defgroup senss_sensor_callbacks Sensor Callbacks
+ * @ingroup senss_sensor
+ */
+
+/**
+ * @brief Sensor API
+ * @addtogroup senss_sensor
+ * @{
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Sensor registration information
+ *
+ */
+struct senss_sensor_register_info {
+
+	/**
+	 * Sensor flags
+	 */
+	uint16_t flags;
+
+	/**
+	 * Sample size in bytes for a single sample of the registered sensor.
+	 * senss runtime need this information for internal buffer allocation.
+	 */
+	uint16_t sample_size;
+
+	/**
+	 * The number of sensor sensitivities
+	 */
+	uint8_t sensitivity_count;
+
+	/**
+	 * Sensor version.
+	 * Version can be used to identify different versions of sensor implementation.
+	 */
+	struct senss_sensor_version version;
+};
+
+/**
+ * @brief Sensor context data structure
+ *
+ */
+struct senss_sensor_ctx {
+
+	/**
+	 * For senss runtime internal private data, sensor should not see and touch
+	 */
+	void *priv_ptr;
+
+	/**
+	 * Pointer to the sensor register information.
+	 */
+	const struct senss_sensor_register_info *register_info;
+
+	/**
+	 * For sensor private context data, registered by sensor with \ref SENSS_SENSOR_DT_DEFINE.
+	 * Sensor could use \ref senss_sensor_get_ctx_data to fetch out this filed with
+	 * struct device.
+	 */
+	void *const sensor_ctx_ptr;
+};
+
+static inline int senss_sensor_dev_init(
+		const struct device *dev)
+{
+	/**
+	 * Nothing need to do in system auto initialization.
+	 * Sensor subsystem runtime will call each sensor instance's initialization
+	 * function via API callback according sensor reporting dependency sequences.
+	 * Sensor subsystem can make sure the depends sensor instances always initialized before
+	 * client sensors.
+	 */
+	return 0;
+}
+
+/**
+ * @brief Macro for define a sensor instance from devicetree node id
+ *
+ * This macro also defined a struct device for this sensor instance, and registered sensors'
+ * private context data, configuration data structure and API.
+ *
+ * \ref senss_init will enumerate all sensor instances from devicetree, and initialize each sensor
+ * instance defined by this macro.
+ *
+ */
+
+#define SENSS_SENSOR_DT_DEFINE(node_id, reg_ptr, ctx_ptr, api_ptr)			\
+	static struct senss_sensor_ctx							\
+		_CONCAT(__senss_sensor_ctx_, Z_DEVICE_DT_DEV_ID(node_id)) = {		\
+			.register_info = reg_ptr,					\
+			.sensor_ctx_ptr = ctx_ptr,					\
+		};									\
+	SENSOR_DEVICE_DT_DEFINE(node_id, senss_sensor_dev_init, NULL,			\
+			&_CONCAT(__senss_sensor_ctx_, Z_DEVICE_DT_DEV_ID(node_id)),	\
+			NULL, APPLICATION, 10, api_ptr)
```

Comment:
> > My understand of option A can save memory because it assumes the application client(s) will access all the sensors from the subsystem and it saves the per sensor client map data. But if an application just need access few sensors of the subsystem, how does the sensor subsystem to allocate per sensor metadata (config data may allocated by user according needs, how about the runtime state data), pre allocate for all sensors will waste of memory.
> 
> If an application doesn't need of the available sensors, then it should disable the sensors it doesn't want to use in the devicetree.

We can config sensors and describe reporting relationships between sensors in devicetree, but we can't describe Application clients in devicetree.

Consider below scenario:   

We have 5 sensors in SENSS (3 physical, 2 virtual), there're two Application clients accessing these sensors **at the same time**.     

As below diagram shows, **solid arrow lines** stand for connection **between sensors**, **dashed arrow lines** stand for connection **between sensor and Application client**, Application client 1 and Application client 2 accessing different sets of sensors: 

![Capture](https://user-images.githubusercontent.com/45843554/230804588-82fd6789-f65b-4aaf-824b-7bc53154283c.PNG)

Yes, we can config these 5 sensors with reporting relationships between them in devicetree (the bluebox in the diagram), but we can't describe how many Application clients and how these clients will access which sensors in devicetree.

Each connection (including between sensors and between sensor and Application client) need memory to store per-connection and per-sensor config data and runtime status data due to multiple clients support and arbitration. 

I think the discussion is focus on **how to save the memory allocation with connection between sensor and Application client**, connection between sensors can be described in devicetree and memory resource can be static allocated at build time, so have no problem. Connection between sensor and Application client is dynamic and can't be described in devicetree at current, @yperess and @MaureenHelm please correct me, if my understanding is wrong. 

And there're two options for this:

- **A: subsystem level connection**
- **B: per sensor connection (this PR)**    

**Option A** can save some memory (only 2 pointer data of 8 bytes for mapping of each sensor, per-sensor config data and runtime status data still can't be saved) for **Application client 1** which need access almost every sensors in SENSS with huge sensors, **this is good for the case that all Application clients need access almost every sensors**, **but it not good for some Application client(s) like Application client 2 which only need access very few sensors at the same time**, since sensor subsystem doesn't know which sensors it will be accessed, open a subsystem level connection means need allocate metadata for all possible accessed sensors, in this case, it will **waste of memory**. 

**Option B** used `senss_open_sensor()/senss_close_sensor()` APIs for Application to explicitly access sensors, SENSS can only allocate necessary memory in `senss_open_sensor()` for real accessed sensors, it's good for the scenarios that's some Application client(s) only need access few sensors while other Application client(s) which running at the same time access almost every sensor in a system with huge sensors.  

**Option A** can only save limited memories for all Application clients need to access almost all sensors in the subsystem, to consider more scenarios, I think **Option B** is more better since it has per-sensor level APIs for memory allocation.  

The problem of **Option B** is it's need dynamic allocate of memory (use malloc in `senss_open_sensor()`) for per-sensor metadata of a connection, but **Option A** also need dynamic allocate of memory (in `senss_open()`) 

For per-sensor metadata of a connection, to avoid of malloc is another problem I think, we're trying best to avoid malloc in SENSS, but for connection between sensor and Application client, at current still not found good static memory allocation solution, especially want to keep concise and easy use APIs. 

Maybe we can provide a parameter for user static defined data structure in `senss_open_sensor()`, SENSS internally use this data structure to save metadata, this can avoid internal malloc, like this:  

```c
SENSS_SENSOR_METADATA_ACC3D(acc0_metadata);
int handle; 
senss_open_sensor(SENSS_SENSOR_TYPE_MOTION_ACCELEROMETER_3D, 0, &acc0_metadata, &handle);

```

But this method need define many macros such as `SENSS_SENSOR_METADATA_XX` by sensor types (different sensor type need different size of metadata), and the API is not concise, so, if possible, I prefer internal malloc, this can keep API more concise and easier to use, since the malloc just happens in initial stage, should not have many impacts. Hi @yperess and @MaureenHelm ,How do you think about it, any suggestions?  Thanks!        

> I think combining configuration parameters into a struct would be more consistent with other Zephyr APIs

I think it's ok to follow common Zephyr APIs style for more consistent. A little problem is there 're two kinds of config parameters: 

- 1: value per client of a sensor
  - report interval, report latency etc 
- B: value per data field/axis of a client of a sensor
  - sensitivity, in feature will have more like offset, range etc. 

so, we can do like this (base on Yuval's suggestion)

```c
struct senss_sensor_config {
  enum senss_sensor_attribute attribute;
  int8_t data_field;
  union {
    int32_t int_val;
    bool bool_val;
    ...
  };
}

int senss_configure_sensor(handle, struct senss_sensor_config *configs, size_t count);
```
For attribute which does not support separated data field, can ignore `data_field` field.